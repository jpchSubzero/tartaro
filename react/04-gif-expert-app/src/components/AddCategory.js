//Snippet -> rafc
import React, { useState } from 'react';

import { PropTypes } from 'prop-types';


//UpperCamelCase (más conocido como PascalCase) para saber que es un componente
// export const AddCategory = () => {

// Recibir función desde el padre y con destructuración asignarle el nombre
export const AddCategory = ({ setCategories }) => {

    // Tarea 2
    //     Recibir la función de agregar categoría
    //     Agregar una categoría al presionar enter

    // Tarea 3
    // Agregar propTypes para requerir la función desde el padre
    // Validar el error cuando no se envía

    //Quitamos el valor inicial, esto dará un error nuevo en consola porque inputValue va a ser undefined y se necesita un valor en <input ... value={inputValue} ... />>
    // const [inputValue, setInputValue] = useState('Valor inicial');

    // Para mostrar el error de valor undefined
    // const [inputValue, setInputValue] = useState();

    // Para corregir el error de undefined que es lo mismo que se hacia en angular, dar un valor inicial
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        // Para visualizar el value desde el target del evento
        // console.log(e.target.value);
        setInputValue(e.target.value);
    }

    // Para manejar presionar enter
    const handleSubmit = (e) => {
        // Para prevenir el comportamiento por defecto del evento
        e.preventDefault();

        console.log('Submit');

        //Invocar la función recibida desde el padre y utilizarla a través del callback porque no tenemos acceso a categories, se puede recibir también las categories pero gracias al callback podemos pedirlas
        // setCategories(categs => [...categs, inputValue]);

        // Agregar validaciones porque acepta vacíos
        if (inputValue.trim()) {
            // Quitar agregar al final para que se vayan agregando al inicio
            // setCategories(categs => [...categs, inputValue]);            

            // Quitar agregar al final para que se vayan agregando al inicio
            setCategories(categs => [inputValue, ...categs]);            

            setInputValue('');
        }
    }

    return (
        // Quitamos <> porque forms ya engloba todos los demás tags
        // <>

        // Agregamos form para poder utilizar el evento submit
        <form onSubmit={ handleSubmit }>
            {/* Para verificar el componente, como AddCategory works!!! en angular */}
            {/* <h2>AddCategory</h2> */}

            {/* Agregar para mostrar el valor */}
            <h1>{ inputValue }</h1>

            {/* Agregar CSS si se quiere */}
            <input 
                type="text" 
                // Produce un error ya que quieres editar un valor inmutable (inputValue)
                value={inputValue}
                // Quita el error pero aún no es editable, debemos crear un manejador para ese evento y allí actualizarlo con el useState (setInputValue)
                // onChange={ (e) => console.log(e) }

                // El evento llama a handleInputChange para actualizar, como es solo una instrucción y recibe e de evento no es necesaria la expresión lambda
                onChange={ handleInputChange }

                // Agregar placeholder después de quitar el valor por defecto
                placeholder="Ingrese una categoría"
            />
        </form>
        // </>
    )
}

AddCategory.propTypes = {
    setCategories:PropTypes.func.isRequired
}