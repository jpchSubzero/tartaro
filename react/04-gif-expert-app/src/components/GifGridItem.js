import React from 'react';

// export const GifGridItem = (props) => {

//Para no usar en el componente props.image enviamos destructurado el objeto
export const GifGridItem = ({id, title, url}) => {

    console.log(title);

    return (
        // <div className="card animate__animated animate__bounce">

        // Agregar animate.css
        <div className="card animate__animated animate__fadeIn">
            {/* Para revisar si lleg√≥ la imagen */}
            {/* { title } */}
            <img src={url} alt={title} />
            <p>{ title }</p>
        </div>
    )
}
