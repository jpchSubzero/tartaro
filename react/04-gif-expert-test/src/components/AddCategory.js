//Snippet -> rafc
import React, { useState } from 'react';

import { PropTypes } from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);

        //Agregar para prueba unitaria
        console.log('Se llamó el método handleInputChange');
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        //Agregar para prueba unitaria
        console.log('Se llamó el método handleSubmit', inputValue);

        if (inputValue.trim().length > 2) {
            setCategories(categs => [inputValue, ...categs]);   
            // setCategories(2412341234);            
            setInputValue('');
        }
    }

    return (
        <form onSubmit={ handleSubmit }>
            <h1>{ inputValue }</h1>

{/* Agregamos el <p> por pruebas unitarias */}
<p>{inputValue }</p>

            <input 
                type="text" 
                value={inputValue}
                onChange={ handleInputChange }
                placeholder="Ingrese una categoría"
            />
        </form>
        // </>
    )
}

AddCategory.propTypes = {
    setCategories:PropTypes.func.isRequired
}