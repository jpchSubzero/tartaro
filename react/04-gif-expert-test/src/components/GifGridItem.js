import React from 'react';

//Agregar propTypes para que envien title y url obligatoriamente
//snapshot -> impt
import PropTypes from 'prop-types';

export const GifGridItem = ({title, url}) => {

    console.log(title);

    return (
        <div className="card animate__animated animate__fadeIn">
            <img src={url} alt={title} />
            <p>{ title }</p>
        </div>
    )
}

//Agregar propTypes para que envien title y url obligatoriamente
GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}

// Tarea 1
// Enzyme
// Enzume to json
// Test para GifGridItem
//     shallow
//     wrapper
//     wrapper .toMatchSnapshot()

// Tarea 2
// Agregar propTypes para que envien title y url obligatoriamente
// Enviar los parametros en la prueba
// Actualizar el snapshot

// Tarea 3
// Validar que el componente tenga la clase animate__fadeIn