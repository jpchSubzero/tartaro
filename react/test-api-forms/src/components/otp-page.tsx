import { useSubscription } from '../hooks/useSubscription';
import { useForm } from '../hooks/useForm';
import { useNavigate } from 'react-router-dom';
import { useIntegration } from '../hooks/useIntegration';
import { useEffect } from 'react';
import IRootState from '../store/store';
import { useSelector } from 'react-redux';
import { useSale } from '../hooks/useSale';

export const OtpPage = () => {
  const { 
    loading, 
    setLoading,
    createOtpHandler, 
    checkOtpHandler, 
    otpValidated,
    otpGenerated
  } = useIntegration();
  const { 
    registerSaleHandler, 
    confirmPaymentHandler, 
    finalizeSaleHandler,
    loading:loadingSale
  } = useSale();

  const navigate = useNavigate();
  const { otp, saleRegister } = useSelector((state:IRootState) => state);
  const [formValues, handleInputChange] = useForm({
    otpInput: "",
  });
  const { otpInput } = formValues;

  const handleValidateSubmit = async (e:any) => {
    e.preventDefault();
    await checkOtpHandler(otpInput);
  };

  const handleRecreateOtpSubmit = async (e:any) => {
    e.preventDefault();
    createOtp();
  };

  const createOtp = async () => {
    await createOtpHandler();
  };
  
  const registerSaleAsync = async () => {
    await registerSaleHandler(saleRegister);
    await confirmPaymentHandler();
    await finalizeSaleHandler();
};
  
  useEffect(() => {
    if (!otpGenerated) {
      createOtp();
    } else {
      if (otp.expiration < new Date()) {
        console.log("otp caducado");
        navigate("/sale");
      }      
    }
  }, []);
  
  useEffect(() => {
    if (otpGenerated) {
      if (otpValidated) {
        console.log('Correcto');
        setLoading(true);
        registerSaleAsync();
      } else {
        console.log('Incorrecto');
      }        
    }
  }, [otpValidated]);
  
  const isValidForm = ():Boolean => {
    return true;
  }

  const renderDefault = () => { 
    return (
      <>
        <form onSubmit={ handleValidateSubmit }>
          <h1>Verificar OTP</h1>  
          <hr />
          <div>
              <input 
                  type="text" 
                  className="form-control"
                  name="otpInput"
                  placeholder="Ingrese la otp"
                  autoComplete="off"
                  value={otpInput}
                  onChange={ handleInputChange}
              ></input>
          </div>

          <button type="submit">Validar</button>

        </form>

        <button onClick={handleRecreateOtpSubmit}>Volver a generar</button>

        <hr />      
      </>
    );
  }

  const renderComponent = () => { 
    return (
      <>
        {renderDefault()}
        <h3><pre>{JSON.stringify(otp, null, 2)}</pre></h3>
      </>
    );
  }

  const renderLoading = () => { 
    return (
      <>
        {loading ? <h1>Cargando...</h1> : <h3><pre>{JSON.stringify(otp, null, 2)}</pre></h3>}
      </>
    );
  }

  if (loading) {
    return renderLoading();    
  } else {
    return renderComponent();        
  }
}
