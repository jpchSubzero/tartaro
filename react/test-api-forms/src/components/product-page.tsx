import { useProduct } from '../hooks/useProduct';
import { IProductFeature } from '../interfaces/products/productFeature.interface';
import { ProductFeatures } from './product.features';

export const ProductPage = () => {
    const { product, loading} = useProduct();
    const productData = product?.result?.$values[0];
    const coverages = productData?.coverages.$values;
    const exclusions = productData?.exclusions.$values;
    const assistances = productData?.assistances.$values;

    const renderComponent = () => { 
        return (
            <>
                <ProductFeatures features={coverages as IProductFeature[]} title="Coberturas" version="0" />
                <hr />
                <ProductFeatures features={exclusions as IProductFeature[]} title="Exclusiones" version="0" />
                <hr />
                <ProductFeatures features={assistances as IProductFeature[]} title="Asistencias" version="0" />
            </>
        );
    }
    
    const renderLoading = () => { 
        return (
            <>
                <h1>Cargando...</h1>
            </>
        );
    }    

    if (loading) {
        return renderLoading();    
    } else {
    return renderComponent();        
    }
}
