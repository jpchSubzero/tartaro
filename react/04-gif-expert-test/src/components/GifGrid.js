// import React, { useState, useEffect } from 'react';
// import { GifGridItem } from './GifGridItem';
// import { getGifs } from '../helpers/getGifs';

import React from 'react';
import { useFetchImages } from '../hooks/useFetchImages';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    //Para explicar el problema de llamar el método getGifs() desde la raíz del componente. Crea un bucle al actualizar count y refrescar el componente. Se utiliza useEffect para ejecutar código en ciertos momentos
    // const [count, setCount] = useState(0);

    //El primer parámetro es una función donde va el código a ejecutar. El segundo parámetro es una lista de dependencias. 
    //[] es que solo se ejecute una vez, como el constructor
    //El warning que empezará a aparecer es porque useEffect detecta que hay el parámetro category y que podría cambiar por lo que sugiere agregarlo como dependencia para que se vuelva a ejecutar el código si category cambia, en este caso no se necesita porque la lógica no va de ese modo. En caso de necesitarse sería [category] como dependencia en el segundo parámetro
    // useEffect(() => {
    //     // Comentar para utilizar el helper getGif
    //     // getGifs();

    //     //Forma larga
    //     // getGifs(category).then((resolve, reject) => {
    //     //     setImages(resolve);
    //     // });

    //     //Forma corta porque getGifs() solo recibe un parámetro y por defecto le pasa resolve
    //     //Luego de esto comentar lo siguiente para implementar el custom hook:
    //     // - useState
    //     // - useEffect
    //     // - div
    //     // import React, { useState, useEffect } from 'react';
    //     // import { GifGridItem } from './GifGridItem';
    //     // import { getGifs } from '../helpers/getGifs';
    //     // Dejar unicamente el import del react
        //     getGifs(category).then(setImages);
    // }, []);

    // //Para manejar las imágenes y luego realizar la tarea
    // const [images, setImages] = useState([]);


    // const getGifs = () => { 

    // //Convertir en función asíncrona
    // const getGifs = async () => { 
    //     // Comentar para agregar la categoría que llega por props luego de las clases
    //     // const url = 'https://api.giphy.com/v1/gifs/search?q=saint+seiya&limit=20&offset=5&rating=g&lang=en&random_id=e826c9fc5c929e0d6c6d423841a282aa&api_key=WzZmmeMbM2fIoV97Bv3ZDXYM6pE1MDa6';

    //     // Agregar la categoría que llega por props y usar encodeURI() para sanitizar el string. También cambiar en AddCategory() para que agregue las nuevas búsquedas al inicio del arreglo
    //     const url = `https://api.giphy.com/v1/gifs/search?q=${ encodeURI(category) }&limit=20&offset=5&rating=g&lang=en&random_id=e826c9fc5c929e0d6c6d423841a282aa&api_key=WzZmmeMbM2fIoV97Bv3ZDXYM6pE1MDa6`;

    //     //Hacer una llamada http
    //     const response = await fetch(url);

    //     //Trae toda la data con envoltorio incluido
    //     // const data = await response.json();

    //     //Destructurar para obtener la data sin envoltorio porque la estructura es obj.data
    //     const { data } = await response.json();

    //     //Usar map para recorrer la data y solo sacar lo que necesitamos
    //     const gifs = data.map(img => {
    //         return {
    //             id: img.id,
    //             title: img.title,
    //             url: img.images?.downsized_medium.url
    //         }
    //     });

    //     setImages(gifs);

    //     console.log(gifs);
    // }

    // Para explicar el problema de llamar el método getGifs() desde la raíz del componente.
    // getGifs();

    //Aquí llamamos al custom hook useFetchImages
    // const state = useFetchImages();
    // console.log(state);

    //Destructurar useFetchImages
    // const { loading, data } = useFetchImages();
    // console.log(loading);
    // console.log(data);

    //Agregar la categoría para que el hook sepa que buscar
    // const { loading, data } = useFetchImages(category);

    // Agregar la referencia images a data para cambiar el nombre
    const { loading, data:images } = useFetchImages(category);


    return (
        // Cambiar por el div para aplicar clases y luego volver a habilitar para poder tener el h3 y el div
        <>

        {/* <h3>{ category }</h3> */}

        {/* Agregar animación */}
        <h3 className="animate__animated animate__fadeIn">{ category }</h3>

        {/* Agregar luego de llamar al custom hook useFetchImages */}
        {/* Agregar el setTimeout en el hook para mostrar como cambia el mensaje cargado */}
        {/* Quitar cuando se use la data */}
        {/* { loading ? 'Cargando...' : 'Data cargada' } */}

        {/* <div className="card-grid"> */}

            {/* Mover fuera del div para que no le afecte el estilo */}
            {/* <h3>{ category }</h3> */}

            {/* Para explicar el problema de llamar el método getGifs() desde la raíz del componente. */}
            {/* <h3>{ count }</h3>
            <button onClick={ () => setCount(count + 1) } >Click</button> */}

            {/* Tarea 4
            Una vez pasados los gifs a través del setImages llenar la lista usando el id como key y el title como valor */}

            {/* Comentar toda la lista ordenada para reemplazar con el componente GifGridItem */}
            {/* <ol> */}
                {/* <li>sfasdfasdf</li> */}
                {
                    // images.map(({id, title}) => {
                    //     if (title) {
                    //         return (<li key={ id }>{ title }</li>);
                    //     }
                    // })
                }
            {/* </ol> */}

            {
                // images.map((image) => {
                //     if (!image.title) {
                //         return '';
                //     }
                //     return (
                //         <GifGridItem 
                //             key={image.id}
                //             // image={ image } 

                //             //Para no usar en el componente props.image enviamos destructurado el objeto
                //             {...image}
                //         />
                //     );                    
                // })
            }

        {/* </div> */}

        {/* Acualizar el div con la data que llega desde el custom hook y agregar un nuevo loader */}
        {/* Este loading usa un operador lógico and y evalúa si ambas son verdaderas y así presenta el mensaje */}
        {/* { loading && <p>Cargando...</p> } */}

        {/* Agregar animación         */}
        { loading && <p className="animate__animated animate__flash">Cargando...</p> }

        <div className="card-grid">
        {
                // Cambiar de data a images mediante la referencia :images
                // data.map((image) => {
                images.map((image) => {
                    if (!image.title) {
                        return '';
                    }
                    return (
                        <GifGridItem 
                            key={image.id}
                            {...image}
                        />
                    );                    
                })
            }
        </div>

        {/* Cambiar por el div para aplicar clases y luego volver a habilitar*/}
        </>
    )
}
