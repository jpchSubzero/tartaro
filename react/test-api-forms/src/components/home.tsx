import { SubscriptionPage } from "./subscription-page";
import { ProductPage } from "./product-page";
import { SalePage } from "./sale-page";
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";
import { Navbar } from "./navbar";
import { Thanks } from "./thanks-page";
import { Error } from "./error-page";
import { OtpPage } from "./otp-page";

export const Home = () => {

    return (
        <>
            <Router>
                <Navbar />
                <Routes>
                    <Route path="/subscription" element={<SubscriptionPage />} />
                    <Route path="/sale" element={<SalePage />} />
                    <Route path="/thanks" element={<Thanks />} />
                    <Route path="/error" element={<Error />} />
                    <Route path="/otp" element={<OtpPage />} />
                    <Route path="/*" element={<ProductPage />} />
                </Routes>
            </Router>
        </>
    )
}
