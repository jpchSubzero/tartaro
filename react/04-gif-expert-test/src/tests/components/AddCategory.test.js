//Importar para que funcione enzyme
import { shallow } from 'enzyme';

//Importar para poder utilizar JSX
import React from 'react';

import { AddCategory } from '../../components/AddCategory';

describe('Pruebas de AddCategory', () => {
    //Creamos un función vacía para poder quitar el error de que el componenete requiere una función como parámetro
    // const setCategories = () => {};

    // Cambiamos la función vacía por su equivalente con jest para poder analizarla
    const setCategories = jest.fn();

    //Mover el wrapper ya que se repite
    // const wrapper = shallow(<AddCategory setCategories={ setCategories }/>);
    let wrapper;

    beforeEach(() => {
        jest.resetAllMocks();

        // Reiniciamos el wrapper, cambiamos const por let y dejamos sin inicializar en la definición. Hay que validar si el intellisence sigue funcionando con el wrapper, si no es así inicializarla para no perder esa ayuda
        wrapper = shallow(<AddCategory setCategories={ setCategories }/>);
    });

    test('Debe mostrarse correctamente', () => {
        // const wrapper = shallow(<AddCategory setCategories={ setCategories }/>);
        expect(wrapper).toMatchSnapshot();
    });    

    test('Validar que cambie la caja de texto', () => {
        // Agregamos un clg para ver que se ejecute el método del componente
        const input = wrapper.find('input');

        //Es igual que en react pero sin el on, onClick, click, onChange, change

        //Nos dará um error de undefined porque el método utiliza el evento y no se está enviando
        // input.simulate('change');

        //Agregamos un evento para superar el error
        const value = 'Hola mundo';

        // Definimos el objeto tipo evento y como en la función referenciamos el target y el value debemos reconstruir la misma estructura
        // Función del componente: setInputValue(e.target.value);
        // Objeto "evento"
        // { 
        //     target: {
        //         //No se especifica value: value ya que tienen el mismo nombre y js sabe que es una asignación
        //         value
        //     }
        // }

        const evento = { 
            target: {
                value
            }
        };
        input.simulate('change', evento);

        // Agregamos el <p> en el componente para validar si se cambia el inputValue
        // Se puede agregar texto, value + "algo", para hacer fallar la prueba
        expect(wrapper.find('p').text().trim()).toBe(value);
    });
    
    test('No debe hacer nada con onSubmit', () => {
        // Cambiamos la función vacía por su equivalente con jest para poder analizarla

        // Agregamos un clg para ver que se ejecute el método del componente

        // Creamos una función vacía porque submit requiere preventDefault
        // const preventDefault = () => {};
        // const evento = { 
        //     preventDefault
        // };

        // Forma corta de definir la función vacía
        const evento = { 
            preventDefault(){}
        };

        wrapper.find('form').simulate('submit', evento);

        // No debe hacer nada ya que la función onSubmit valida cuando el inputValue no tiene nada (length > 2) 
        // Si no presenta ya ayuda importar @testing-library/jest-dom
        // Este expect nos devolverá un error de que se ejecutó una vez setCategories(), esto es porque en la prueba anterior se llama a input.simulate('change', evento) y este cambia el valor de inputValue, por lo tanto su longitud ya no es 0. Para ver mejor agregar inputValue al clg de handleSubmit() del componente. Para solucionar el problema se debe reinicializar la prueba con beforeEach()
        expect(setCategories).not.toHaveBeenCalled();
    });

    // Tarea 1
    test('Debe llamar setCategories y limpiar la caja de texto', () => {
        const value = 'Hola mundo';

        // Simular inputChanges
        wrapper.find('input').simulate('change', { target: { value: value} });

        // Simular el submit
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        // Debe haberse llamado setCategories. Se puede agregar el número de veces que se debía haber ejecutado con el evaludador toHaveBeenCalledTimes(<cantidad>)
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);

        // El value del input debe ser vacío
        expect(wrapper.find('input').text().trim()).toBe('');

        // Las validaciones de la tarea funcionan pero no validan si se está usando una función como parámetro en setCategories, para probar hay que cambiar la función setCategories(categs => [inputValue, ...categs]);  del componente por setCategories(234124);  es decir, por una función que recibe números, esto causa que la prueba sea correcta pero en contexto estaría mal. Para probar mejor se puede indicar que el parámetro de la función debe ser una función con expect.any(Function) y el evaluador toHaveBeenCalledWith()
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
    })
    
})

// Tarea 1
// Simular inputChanges
// Simular el submit
// Debe haberse llamado setCategories
// El value del input debe ser vacío