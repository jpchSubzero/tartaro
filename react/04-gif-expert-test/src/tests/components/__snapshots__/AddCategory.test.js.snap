// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de AddCategory Debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <h1 />
  <p />
  <input
    onChange={[Function]}
    placeholder="Ingrese una categorÃ­a"
    type="text"
    value=""
  />
</form>
`;
